src/Types/ListType.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace ILIAS\Types;
+
+/**
+ * Class ListType
+ *
+ * @package ILIAS\Types
+ *
+ * @author  Oskar Truffer <ot@studer-raimann.ch>
+ *
+ * Describes a list of a certain Type.
+ *
+ * Attention:
+ * isExtensionOf behaves Covariant!
+ * new ListType(Dog::class).isExtensionOf(new ListType(Animal::class)) == true!
+ *
+ * See: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
+ */
+class ListType implements Type, Ancestors {
+
+	/** @var Type */
+	protected $type;
+
+
+	/**
+	 * SingleType constructor.
+	 *
+	 * @param $fullyQualifiedClassName string|Type Give a Value Type or a Type that will be wrapped in a single type.
+	 */
+	public function __construct($fullyQualifiedClassName) {
+		if (!is_a($fullyQualifiedClassName, Type::class)) {
+			$fullyQualifiedClassName = new SingleType($fullyQualifiedClassName);
+		}
+		$this->type = $fullyQualifiedClassName;
+	}
+
+
+	/**
+	 * @inheritdoc
+	 */
+	function __toString() {
+		return "[" . $this->type . "]";
+	}
+
+
+	/**
+	 * Is this type a subtype of $type. Not strict! x->isExtensionOf(x) == true.
+	 *
+	 * Attention: [Dog].isExtensionOf([Animal]) == true. In other words:
+	 * The isExtensionOf behaves Covariant!
+	 *
+	 * If you are familiar with e.g. Java Generics ?.isExtensionOf(x) behaves the same as <? extends x>.
```

Comment:
Hi Richard

Notice: In the following if you read A<'T> you can ignore the tick just before the T. It's only there to prevent Github from sanitizing the whole <....> part.

> If I have any two types A and B the only things I could do with them are defined via the interface Type. If A.isExtensionOf(B) === true: What does it tell me? How does it help? I certainly cannot treat the A like a B, because this exactly breaks down in the list case. Thus, if I have a background task (or other thingy) acting on a B, I cannot feed an A into it.

I think I fully understand your point now. My arguments were based too much on the list case and not on this generic approach.

> What is your understanding of covariant and contravariant? As far as I understand, List is neither covariant nor contravariant, hence the problem. pop is covariant. If I have a list that contains As and pop values from it, I could treat the As as Bs if A.isExtensionOf(B). push on the other hand is contravariant. If I have a List of Bs I could push As into it if A.isExtensionOf(B), which is exactly the other way round. If I had a list of As I cannot push Bs into it, even if A.isExtensionOf(B), because an A is somehow more than a B.

Yes, I agree with your definition of co- and contravariant as well as your point here.

To sum up: A (List) that depends on some generic T (Animal) can never implement isSubclassOf with respect to the generic T because methods of A might act co- or contravariant concerning T.

See: 

List<'Animal>.isSubtypeOf(List<'Dog>) == True => List<'Animal>.pop() should give us a Dog. => Contradiction!

List<'Dog>.isSubtypeOf(List<'Animal>) == True => List<'Dog>.push(...) should accept an Animal. => Contradiction!

This means that: A<'T>.isSubtypeOf(A<'U>) == False; For all T !=U. (This was a proof by example but should hold for the general case).

Okay then what about A<? extends T> case? Or in our case List<? extends Animal>. This generalizes the generic to anything that extends Animal. So let's try it:

List<'Dog>.isSubtypeOf(List<? extends Animal>) == True. List<? extends Animal>.push(...) accepts a Cat as it also extends animal. But List<'Dog> does not accept a Cat thus isSubtype == False => Contradiction.

Thus the <? extends..> does not solve this Problem. The only Problem that this solves is that it marks the input as "dangerous". So if you have a method that accepts x of Type A<? extends T> (implemented as mentioned above) you CANNOT invoke any contravariant methods on x.

We have no notion on whether a method acts co- or contravariant on its input or output and I don't want to introduce any, thus we should only allow A<'T>.isSubtypeOf(A<'U>) == true; iff T==U. But then a job that can be invoked on a list of ids cannot be invoked on a list of user ids and this would probably be a nice use-case.

Do you have any solution?

> And also nice to have a discussion on this level here.
It certainly is! Even though it feels like there should be a go-to solution. Wikipedia tells that most implementations use just covariant-y behaviour for our isSubtypeOf/isExtensionOf even though this might be unsound but easier to reason about for programmers. No idea what "most" means here.

Best
Oskar